javaagent-extensions/gcp-auth/src/main/java/com/google/cloud/opentelemetry/extension/auth/GcpAuthAutoConfigurationCustomizerProvider.java

```diff
@@ -0,0 +1,135 @@
+public class GcpAuthAutoConfigurationCustomizerProvider
+    implements AutoConfigurationCustomizerProvider {
+
+  private static final String QUOTA_USER_PROJECT_HEADER = "X-Goog-User-Project";
+  private static final String GCP_USER_PROJECT_ID_KEY = "gcp.project_id";
+
+  /**
+   * Customizes the provided {@link AutoConfigurationCustomizer}.
+   *
+   * <p>This method attempts to retrieve Google Application Default Credentials (ADC) and performs
+   * the following: - Adds authorization headers to the configured {@link SpanExporter} based on the
+   * retrieved credentials. - Adds default properties for OTLP endpoint and resource attributes for
+   * GCP integration.
+   *
+   * @param autoConfiguration the AutoConfigurationCustomizer to customize.
+   * @throws GoogleAuthException if there's an error retrieving Google Application Default
+   *     Credentials.
+   * @throws io.opentelemetry.sdk.autoconfigure.spi.ConfigurationException if required options are
+   *     not configured through environment variables or system properties.
+   */
+  @Override
+  public void customize(AutoConfigurationCustomizer autoConfiguration) {
+    try {
+      GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
+      autoConfiguration
+          .addSpanExporterCustomizer(
+              (exporter, configProperties) -> addAuthorizationHeaders(exporter, credentials))
+          .addPropertiesSupplier(this::getRequiredProperties)
+          .addResourceCustomizer(this::customizeResource);
+    } catch (IOException e) {
+      throw new GoogleAuthException(Reason.FAILED_ADC_RETRIEVAL, e);
+    }
+  }
+
+  @Override
+  public int order() {
+    return Integer.MAX_VALUE - 1;
+  }
+
+  // Adds authorization headers to the calls made by the OtlpGrpcSpanExporter and
+  // OtlpHttpSpanExporter.
+  private SpanExporter addAuthorizationHeaders(
+      SpanExporter exporter, GoogleCredentials credentials) {
+    if (exporter instanceof OtlpHttpSpanExporter) {
+      OtlpHttpSpanExporterBuilder builder =
+          ((OtlpHttpSpanExporter) exporter)
+              .toBuilder().setHeaders(() -> getRequiredHeaderMap(credentials));
+      return builder.build();
+    } else if (exporter instanceof OtlpGrpcSpanExporter) {
+      OtlpGrpcSpanExporterBuilder builder =
+          ((OtlpGrpcSpanExporter) exporter)
+              .toBuilder().setHeaders(() -> getRequiredHeaderMap(credentials));
+      return builder.build();
+    }
+    return exporter;
+  }
+
+  private Map<String, String> getRequiredHeaderMap(GoogleCredentials credentials) {
+    Map<String, String> gcpHeaders = new HashMap<>();
+    try {
+      credentials.refreshIfExpired();
+    } catch (IOException e) {
+      throw new GoogleAuthException(Reason.FAILED_ADC_REFRESH, e);
+    }
+    gcpHeaders.put(QUOTA_USER_PROJECT_HEADER, credentials.getQuotaProjectId());
+    gcpHeaders.put("Authorization", "Bearer " + credentials.getAccessToken().getTokenValue());
```

Comment:
Since the map is just made once, it seems like this won't support refresh?

Reply:
No, the function `getRequiredHeaderMap` is invoked through a header supplier which means it gets recreated during export.

There might be some performance concerns about recreating Map on every export, which I intend to discuss in the SIG tomorrow.